File: templates/templates/ruby/tests/secondary/auth_test.rb
# frozen_string_literal: true
# typed: true

require_relative "../lib/alphabetically_early"
require_relative "common_helper_test"

require "minitest/autorun"
require "minitest/focus"
require "rack"

module AlphabeticallyEarly
  class TestAuth < Minitest::Test
    def setup
      @sdk = AlphabeticallyEarly::SDK.new(server_url: CommonHelpers.httpbin_url)
    end

    def test_no_auth
      record_test("auth-hoisted-no-auth-retained")

      res = @sdk.auth.no_auth

      assert_nil(res)
    end

    def test_basic_auth
      record_test("auth-hoisted-basic-auth")

      @sdk = AlphabeticallyEarly::SDK.new(
        server_url: CommonHelpers.httpbin_url,
        security: Models::Shared::Security.new(
          username: "testUser",
          password: "testPass"
        )
      )

      res = @sdk.auth.basic_auth(
        passwd: "testPass",
        user: "testUser"
      )

      refute_nil(res)
      assert_equal(res.authenticated, true)
    end

    def test_multiple_mixed_options_auth
      record_test("auth-hoisted-operation-auth-retained")

      res = @sdk.auth_new.multiple_mixed_options_auth(
        security: Models::Operations::MultipleMixedOptionsAuthSecurity.new(
          basic_auth: Models::Shared::SchemeBasicAuth.new(
            username: "testUser",
            password: "testPass"
          )
        ),
        request: Models::Shared::AuthServiceRequestBody.new(
          basic_auth: Models::Shared::BasicAuth.new(
            username: "testUser",
            password: "testPass"
          )
        )
      )
      assert_nil(res)
    end

    def test_custom_security_scheme_app_id
      record_test("auth-custom-security-scheme-app-id")

      res = @sdk.auth_new.custom_scheme_app_id(
        security: Models::Operations::CustomSchemeAppIdSecurity.new(
          app_id: "testAppID",
          secret: "testSecret"
        )
      )
      assert_nil(res)
    end
  end
end

File: templates/templates/php/Tests/secondary/AuthTest.php
declare(strict_types=1);

namespace alphabetically\early\Tests;

use alphabetically\early\Tests\CommonHelpers;
use alphabetically\early\Tests\RequestRecorderClient;
use alphabetically\early\Tests\RequestLogEntry;
use function alphabetically\early\Tests\createRequestRecorderClient;
use PHPUnit\Framework\TestCase;

final class AuthTest extends TestCase
{
    public function testNoAuth(): void
    {
        CommonHelpers::recordTest('auth-hoisted-no-auth-retained');

        $sdk = \alphabetically\early\SDK::builder()->setServerUrl(CommonHelpers::getHttpBinUrl())->build();

        $this->assertInstanceOf(\alphabetically\early\SDK::class, $sdk);

        $response = $sdk->auth->noAuth();
    }

    public function testBasicAuth(): void
    {
        CommonHelpers::recordTest('auth-hoisted-basic-auth');

        $security = new \alphabetically\early\Models\Shared\Security();
        $security->username = 'testUser';
        $security->password = 'testPass';

        $sdk = \alphabetically\early\SDK::builder()->setServerUrl(CommonHelpers::getHttpBinUrl())->setSecurity($security)->build();

        $this->assertInstanceOf(\alphabetically\early\SDK::class, $sdk);

        $response = $sdk->auth->basicAuth('testPass', 'testUser');

        $this->assertNotNull($response);
        $this->assertTrue($response->authenticated);
    }

    public function testMultipleMixedOptionsAuth(): void
    {
        CommonHelpers::recordTest('auth-hoisted-operation-auth-retained');
        $sdk = \alphabetically\early\SDK::builder()->setServerUrl(CommonHelpers::getHttpBinUrl())->build();

        $this->assertInstanceOf(\alphabetically\early\SDK::class, $sdk);

        $request = new \alphabetically\early\Models\Shared\AuthServiceRequestBody();
        $request->basicAuth = new \alphabetically\early\Models\Shared\AuthServiceRequestBodyBasicAuth(
            username: 'testUser',
            password: 'testPass',
        );

        $security = new \alphabetically\early\Models\Operations\MultipleMixedOptionsAuthSecurity();
        $security->basicAuth = new \alphabetically\early\Models\Shared\SchemeBasicAuth(
            username: 'testUser',
            password: 'testPass',
        );

        $sdk->authNew->multipleMixedOptionsAuth($security, $request);
    }

    public function testOperationLevelOauth2(): void
    {
        CommonHelpers::recordTest('auth-operation-level-oauth2');

        $log = [];
        $client = createRequestRecorderClient($log);
        $sdk = \alphabetically\early\SDK::builder()->setServerUrl(CommonHelpers::getHttpBinUrl())
            ->setClient($client)
            ->build();
        $clientSecret = 'supersecret-'.CommonHelpers::randSequence(10);

        // A token should be requested with 'read', 'write' and 'erase' scopes.
        $sdk->hooks->authenticatedRequest(
            new \alphabetically\early\Models\Operations\AuthenticatedRequestSecurity(
                audience: '',
                clientID: 'speakeasy-sdks',
                clientSecret: $clientSecret,
            )
        );

        // This operation requires 'read' and 'write' scopes.
        // The same token should be reused since [read, write, erase] is a superset of [read, write].
        $sdk->hooks->authenticatedRequestUnflattened(
            new \alphabetically\early\Models\Operations\AuthenticatedRequestUnflattenedSecurity(
                clientCredentials: new \alphabetically\early\Models\Shared\SchemeClientCredentials(
                    audience: '',
                    clientID: 'speakeasy-sdks',
                    clientSecret: $clientSecret,
                )
            )
        );

        // Verify that only a single token was requested
        $tokenRequests = array_filter($log, function($entry) {
            return str_contains($entry->requestBody ?? '', 'grant_type=client_credentials');
        });
        $this->assertCount(1, $tokenRequests, 'Expected only a single token request');
        $firstRequest = reset($tokenRequests);
        $this->assertStringContainsString('scope=read+write+erase', $firstRequest->requestBody ?? '');
    }

    public function testCustomSecuritySchemeAddId(): void
    {
        CommonHelpers::recordTest('auth-custom-security-scheme-app-id');

        $sdk = \alphabetically\early\SDK::builder()->setServerUrl(CommonHelpers::getHttpBinUrl())->build();

        $this->assertInstanceOf(\alphabetically\early\SDK::class, $sdk);

        $response = $sdk->authNew->customSchemeAppId(
            security: new \alphabetically\early\Models\Operations\CustomSchemeAppIdSecurity(
                appId: 'testAppID',
                secret: 'testSecret'
            )
        );
    }
}
